jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.registry;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpResponseException;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.cloud.tools.jib.http.Authorization;
+import com.google.cloud.tools.jib.http.Connection;
+import com.google.cloud.tools.jib.http.Request;
+import com.google.cloud.tools.jib.http.Response;
+import com.google.cloud.tools.jib.json.JsonTemplateMapper;
+import com.google.cloud.tools.jib.registry.json.ErrorEntryTemplate;
+import com.google.cloud.tools.jib.registry.json.ErrorResponseTemplate;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLPeerUnverifiedException;
+import org.apache.http.NoHttpResponseException;
+
+/** Makes requests to a registry endpoint. */
+class RegistryEndpointCaller<T> {
+
+  private static final String DEFAULT_PROTOCOL = "https";
+
+  /** Maintains the state of a request. This is used to retry requests with different parameters. */
+  private static class RequestState {
+
+    @Nullable private final Authorization authorization;
+    private final URL url;
+
+    /**
+     * @param authorization authentication credentials
+     * @param url the endpoint URL to call, or {@code null} to use default from {@code
+     *     registryEndpointProvider}
+     */
+    private RequestState(@Nullable Authorization authorization, URL url) {
+      this.authorization = authorization;
+      this.url = url;
+    }
+  }
+
+  private final RequestState initialRequestState;
+  private final String userAgent;
+  private final RegistryEndpointProvider<T> registryEndpointProvider;
+  private final RegistryEndpointProperties registryEndpointProperties;
+
+  /**
+   * Constructs with parameters for making the request.
+   *
+   * @param userAgent {@code User-Agent} header to send with the request
+   * @param apiRouteBase the endpoint's API root, without the protocol
+   * @param registryEndpointProvider the {@link RegistryEndpointProvider} to the endpoint
+   * @param authorization optional authentication credentials to use
+   * @param registryEndpointProperties properties of the registry endpoint request
+   */
```

Comment:
Need `@throws` to avoid more javadoc warnings

Reply:
I'll merge your javadoc PR so that the build fails if I don't fix these.

Reply:
maybe fix these for now? before we submit?